File: webapp/static/physics/track.js
// JS port of f1sim/track.py: resample a raw centerline to uniform arc-length
// spacing, and compute curvature/path-length for an offset (racing-line)
// path. Mirrors the Python implementation function-for-function.

function resampleTrack(trackDict, spacing = 2.0) {
  const closed = trackDict.closed !== false;
  let pts = trackDict.points.map((p) => [p[0], p[1]]);
  const rawWidth = trackDict.width ?? 12.0;
  let w = Array.isArray(rawWidth) ? rawWidth.slice() : new Array(pts.length).fill(rawWidth);

  if (closed) {
    pts = [...pts, pts[0]];
    w = [...w, w[0]];
  }

  const cum = [0];
  for (let i = 1; i < pts.length; i++) {
    const dx = pts[i][0] - pts[i - 1][0];
    const dy = pts[i][1] - pts[i - 1][1];
    cum.push(cum[i - 1] + Math.hypot(dx, dy));
  }
  const totalLen = cum[cum.length - 1];
  const nSamples = Math.max(Math.floor(totalLen / spacing), 10);

  const sNew = [];
  if (closed) {
    for (let i = 0; i < nSamples; i++) sNew.push((i / nSamples) * totalLen);
  } else {
    for (let i = 0; i < nSamples; i++) sNew.push((i / (nSamples - 1)) * totalLen);
  }

  function interpAlong(cumArr, valsArr, sArr) {
    const out = new Float64Array(sArr.length);
    let idx = 0;
    for (let k = 0; k < sArr.length; k++) {
      const s = sArr[k];
      while (idx < cumArr.length - 2 && cumArr[idx + 1] < s) idx++;
      const t = (s - cumArr[idx]) / (cumArr[idx + 1] - cumArr[idx] || 1);
      out[k] = valsArr[idx] + t * (valsArr[idx + 1] - valsArr[idx]);
    }
    return out;
  }

  const xs = pts.map((p) => p[0]);
  const ys = pts.map((p) => p[1]);
  const x = interpAlong(cum, xs, sNew);
  const y = interpAlong(cum, ys, sNew);
  const wNew = interpAlong(cum, w, sNew);

  const points = [];
  for (let i = 0; i < nSamples; i++) points.push([x[i], y[i]]);

  return new ResampledTrack(points, wNew, closed, sNew);
}

class ResampledTrack {
  constructor(points, width, closed, s) {
    this.points = points;
    this.width = width;
    this.closed = closed;
    this.s = s;
    this.n = points.length;
    this._computeFrames();
  }

  _computeFrames() {
    const n = this.n;
    const tangent = new Array(n);
    const normal = new Array(n);
    for (let i = 0; i < n; i++) {
      let prev, next;
      if (this.closed) {
        prev = this.points[(i - 1 + n) % n];
        next = this.points[(i + 1) % n];
      } else {
        prev = this.points[Math.max(i - 1, 0)];
        next = this.points[Math.min(i + 1, n - 1)];
      }
      let dx = next[0] - prev[0], dy = next[1] - prev[1];
      const norm = Math.hypot(dx, dy) || 1e-9;
      dx /= norm; dy /= norm;
      tangent[i] = [dx, dy];
      normal[i] = [-dy, dx]; // left-hand normal (90deg CCW)
    }
    this.tangent = tangent;
    this.normal = normal;
  }

  /** Curvature of the path offset by alpha (fraction of half-width, -1..1). */
  curvatureOfOffset(alpha) {
    const n = this.n;
    const path = new Array(n);
    for (let i = 0; i < n; i++) {
      const off = (alpha[i] * this.width[i]) / 2.0;
      path[i] = [this.points[i][0] + off * this.normal[i][0], this.points[i][1] + off * this.normal[i][1]];
    }

    const kappa = new Float64Array(n);
    for (let i = 0; i < n; i++) {
      let prev, next;
      if (this.closed) {
        prev = path[(i - 1 + n) % n];
        next = path[(i + 1) % n];
      } else {
        prev = path[Math.max(i - 1, 0)];
        next = path[Math.min(i + 1, n - 1)];
      }
      const d1x = next[0] - path[i][0], d1y = next[1] - path[i][1];
      const d2x = path[i][0] - prev[0], d2y = path[i][1] - prev[1];
      let ds1 = Math.hypot(d1x, d1y) || 1e-9;
      let ds2 = Math.hypot(d2x, d2y) || 1e-9;
      const cross = d2x * d1y - d2y * d1x;
      kappa[i] = (2 * cross) / (ds1 * ds2 * (ds1 + ds2));
    }
    return { kappa, path };
  }

  pathLength(alpha) {
    const { path } = this.curvatureOfOffset(alpha);
    const n = this.n;
    const seg = new Float64Array(n);
    for (let i = 0; i < n; i++) {
      const next = this.closed ? path[(i + 1) % n] : path[Math.min(i + 1, n - 1)];
      seg[i] = Math.hypot(next[0] - path[i][0], next[1] - path[i][1]);
    }
    return seg;
  }
}

File: f1sim/track.py
"""Track loading and geometry: centerline, resampling, curvature, normals."""
import json
import numpy as np


class Track:
    def __init__(self, name, points, width, closed=True):
        """points: (N,2) raw centerline waypoints in meters.
        width: scalar or (N,) array of track width in meters.
        closed: whether the track loops back to the start."""
        self.name = name
        self.closed = closed
        self.raw_points = np.asarray(points, dtype=float)
        n = len(self.raw_points)
        self.raw_width = np.full(n, width, dtype=float) if np.isscalar(width) else np.asarray(width, dtype=float)

    @classmethod
    def from_dict(cls, data):
        return cls(
            name=data.get("name", "unnamed"),
            points=data["points"],
            width=data.get("width", 12.0),
            closed=data.get("closed", True),
        )

    @classmethod
    def from_json(cls, path):
        with open(path, "r") as f:
            data = json.load(f)
        return cls.from_dict(data)

    def resample(self, spacing=2.0):
        """Resample centerline to ~uniform arc-length spacing. Returns a
        RacingLine-ready set of arrays: points, width, tangent, normal, s."""
        pts = self.raw_points
        w = self.raw_width
        if self.closed:
            pts = np.vstack([pts, pts[0]])
            w = np.append(w, w[0])

        seg = np.diff(pts, axis=0)
        seg_len = np.linalg.norm(seg, axis=1)
        cum = np.concatenate([[0], np.cumsum(seg_len)])
        total_len = cum[-1]

        n_samples = max(int(total_len / spacing), 10)
        s_new = np.linspace(0, total_len, n_samples, endpoint=not self.closed)

        x = np.interp(s_new, cum, pts[:, 0])
        y = np.interp(s_new, cum, pts[:, 1])
        w_new = np.interp(s_new, cum, w)

        return ResampledTrack(np.column_stack([x, y]), w_new, self.closed, s_new)


class ResampledTrack:
    """Uniformly-sampled closed/open centerline with derived geometry."""

    def __init__(self, points, width, closed, s):
        self.points = points          # (N,2)
        self.width = width            # (N,)
        self.closed = closed
        self.s = s                    # cumulative arc length (N,)
        self.n = len(points)
        self.ds = np.gradient(s) if not closed else self._closed_ds()
        self._compute_frames()

    def _closed_ds(self):
        s = self.s
        ds = np.diff(s, append=s[0] + (s[-1] - s[-2]) + (s[1] - s[0]))
        return ds

    def _compute_frames(self):
        pts = self.points
        if self.closed:
            nxt = np.roll(pts, -1, axis=0)
            prv = np.roll(pts, 1, axis=0)
        else:
            nxt = np.vstack([pts[1:], pts[-1:]])
            prv = np.vstack([pts[:1], pts[:-1]])

        tangent = nxt - prv
        norm = np.linalg.norm(tangent, axis=1, keepdims=True)
        norm[norm == 0] = 1e-9
        tangent = tangent / norm
        # left-hand normal (90 deg CCW)
        normal = np.column_stack([-tangent[:, 1], tangent[:, 0]])

        self.tangent = tangent
        self.normal = normal

    def curvature_of_offset(self, alpha):
        """Curvature of the path offset by alpha (fraction of half-width,
        -1..1, +1 = full offset toward the left normal) at each sample.
        Computed via finite differences on the offset path itself."""
        offset = (alpha * self.width / 2.0)[:, None] * self.normal
        path = self.points + offset

        if self.closed:
            p_prev = np.roll(path, 1, axis=0)
            p_next = np.roll(path, -1, axis=0)
        else:
            p_prev = np.vstack([path[:1], path[:-1]])
            p_next = np.vstack([path[1:], path[-1:]])

        d1 = p_next - path
        d2 = path - p_prev
        ds1 = np.linalg.norm(d1, axis=1)
        ds2 = np.linalg.norm(d2, axis=1)
        ds1[ds1 == 0] = 1e-9
        ds2[ds2 == 0] = 1e-9

        cross = d2[:, 0] * d1[:, 1] - d2[:, 1] * d1[:, 0]
        kappa = 2 * cross / (ds1 * ds2 * (ds1 + ds2))
        return kappa, path

    def path_length(self, alpha):
        _, path = self.curvature_of_offset(alpha)
        if self.closed:
            nxt = np.roll(path, -1, axis=0)
        else:
            nxt = np.vstack([path[1:], path[-1:]])
        seg = np.linalg.norm(nxt - path, axis=1)
        return seg
